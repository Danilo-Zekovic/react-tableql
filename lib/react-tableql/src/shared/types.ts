export interface TableQLThemeAdvanced {
  header?: {
    color?: string
    background?: string
  }
  rows?: {
    color?: string
    background?: string
    hover?: string
    stripes?: {
      color?: string
      background?: string
      hover?: string
    }
  }
  grid?: {
    horizontal?: string
    vertical?: string
    frame?: string
  }
}

export interface TableQLThemeBasic {
  primary?: string // main color for the table, header and colored striped rows
  secondary?: string // rows that are not colored in striped scenario or all rows in non striped setup
  primaryFontColor?: string // font color for header and striped rows
  secondaryFontColor?: string // color for rows
  striped?: boolean // should the table rows be striped
  grid?: string // grid line color
}

export type TableQLTheme = TableQLThemeAdvanced | TableQLThemeBasic

export type Data = object[] | object | []
export interface ColumnConfig {
  id: string
  label?: string
  component?: (data: unknown) => void
  customColumn?: boolean
  headerStyle?: string
  nodeStyle?: string | ((data: unknown) => void)
  sort?: boolean | ((newOrder: unknown, property: string) => void)
}

export type Column = string | ColumnConfig
