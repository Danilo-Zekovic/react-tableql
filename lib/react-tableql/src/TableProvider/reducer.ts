import { Action, State } from './types'

export const tableReducer = (
  state: State,
  { type, payload }: Action,
): State => {
  // if (!theme) {
  //   return { theme: { ...state.theme } }
  // }

  switch (type) {
    case 'setTheme': {
      if (!payload.theme) {
        // TODO consider maybe throwing error
        // throw new Error(`Theme not passed.`)
        return { ...state }
      }
      return { ...state, theme: { ...payload.theme } }
    }
    case 'setData': {
      return { ...state, data: [...payload.data] }
    }

    case 'setDataKeys': {
      return { ...state, dataKeys: [...payload.dataKeys] }
    }
    default: {
      throw new Error(`Unhandled action type: ${type}`)
    }
  }
}
