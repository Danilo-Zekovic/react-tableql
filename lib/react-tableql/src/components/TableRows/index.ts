import TableRows from './TableRows'

export { TableRows }
