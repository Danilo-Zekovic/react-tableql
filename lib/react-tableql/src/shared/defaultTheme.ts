import { TableQLProviderTheme } from '../TableQLProvider/types'

export const defaultTheme: TableQLProviderTheme = {
  header: {
    color: '#000',
    background: 'transparent', //'#fff',
  },
  rows: {
    color: '#000',
    background: 'transparent', //'#fff',
    hover: '#f1f3f5',
    stripes: {
      color: '#000',
      background: 'transparent', //'#fff',
      hover: '#f1f3f5',
    },
  },
  grid: {
    horizontal: '#dee2e6',
    vertical: '#dee2e6',
    frame: '#dee2e6',
  },
}
