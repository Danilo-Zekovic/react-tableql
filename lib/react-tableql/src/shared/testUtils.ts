export const MOCK_DATA = [
  {
    author: 'Austen, Jane',
    title: 'Northanger Abbey',
    id: '63a32284072b2e31190bd81d',
    edition: 'Penguin',
    price: 18.2,
    yearWritten: 1814,
  },
  {
    author: 'Tolstoy, Leo',
    title: 'War and Peace',
    id: '63a32284072b2e31190bd81e',
    edition: 'Penguin',
    price: 12.7,
    yearWritten: 1865,
  },
  {
    author: 'Tolstoy, Leo',
    title: 'Anna Karenina',
    id: '63a32284072b2e31190bd81f',
    edition: 'Penguin',
    price: 13.5,
    yearWritten: 1875,
  },
  {
    author: 'Woolf, Virginia',
    title: 'Mrs. Dalloway',
    id: '63a32284072b2e31190bd820',
    edition: 'Harcourt Brace',
    price: 25,
    yearWritten: 1925,
  },
  {
    author: 'Cunnningham, Michael',
    title: 'The Hours',
    id: '63a32284072b2e31190bd821',
    edition: 'Harcourt Brace',
    price: 12.35,
    yearWritten: 1999,
  },
  {
    author: 'Twain, Mark',
    title: 'Huckleberry Finn',
    id: '63a32284072b2e31190bd822',
    edition: 'Penguin',
    price: 5.76,
    yearWritten: 1865,
  },
  {
    author: 'Dickens, Charles',
    title: 'Bleak House',
    id: '63a32284072b2e31190bd823',
    edition: 'Random House',
    price: 5.75,
    yearWritten: 1870,
  },
  {
    author: 'Twain, Mark',
    title: 'Tom Sawyer',
    id: '63a32284072b2e31190bd824',
    edition: 'Random House',
    price: 7.75,
    yearWritten: 1862,
  },
  {
    author: 'Woolf, Virginia',
    title: "A Room of One's Own",
    id: '63a32284072b2e31190bd825',
    edition: 'Penguin',
    price: 29,
    yearWritten: 1922,
  },
  {
    author: 'Rowling, J.K.',
    title: 'Harry Potter',
    id: '63a32284072b2e31190bd826',
    edition: 'Harcourt Brace',
    price: 19.95,
    yearWritten: 2000,
  },
  {
    author: 'Marquez',
    title: 'One Hundred Years of Solitude',
    id: '63a32284072b2e31190bd827',
    edition: 'Harper  Perennial',
    price: 14,
    yearWritten: 1967,
  },
  {
    author: 'Shakespeare',
    title: 'Hamlet, Prince of Denmark',
    id: '63a32284072b2e31190bd828',
    edition: 'Signet  Classics',
    price: 7.95,
    yearWritten: 1603,
  },
  {
    author: 'Tolkien, J.R.',
    title: 'Lord of the Rings',
    id: '63a32284072b2e31190bd829',
    edition: 'Penguin',
    price: 27.45,
    yearWritten: 1937,
  },
]

export const MOCK_DATA_KEYS = Object.keys(MOCK_DATA[0])
