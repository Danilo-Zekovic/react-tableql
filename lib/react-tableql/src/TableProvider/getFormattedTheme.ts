import { defaultTheme } from '../shared/defaultTheme'
import {
  TableQLTheme,
  TableQLThemeAdvanced,
  TableQLThemeBasic,
} from '../shared/types'
import { TableQLProviderTheme } from '../TableQLProvider/types'

const isBasicTheme = (theme: TableQLTheme): theme is TableQLThemeBasic => {
  if (
    (theme as TableQLThemeBasic).primary ||
    (theme as TableQLThemeBasic).secondary ||
    (theme as TableQLThemeBasic).primaryFontColor ||
    (theme as TableQLThemeBasic).secondaryFontColor ||
    (theme as TableQLThemeBasic).striped ||
    (theme as TableQLThemeBasic).primary
  ) {
    return true
  }
  return false
}

const formattedBasicTheme = ({
  primary,
  primaryFontColor,
  secondaryFontColor,
  secondary,
  grid,
}: TableQLThemeBasic): TableQLProviderTheme => {
  const {
    header,
    rows,
    grid: { horizontal, vertical },
  } = defaultTheme

  return {
    header: {
      color: primaryFontColor || header.color,
      background: primary || header.background,
    },
    rows: {
      color: secondaryFontColor || rows.color,
      background: secondary || rows.background,
      hover: rows.hover,
      stripes: {
        color: primaryFontColor || rows.stripes.color,
        background: primaryFontColor || rows.stripes.background,
        hover: rows.stripes.hover,
      },
    },
    grid: {
      horizontal: grid || horizontal,
      vertical: grid || vertical,
    },
  }
}

const formattedAdvancedTheme = ({
  header,
  rows,
  grid,
}: TableQLThemeAdvanced): TableQLProviderTheme => {
  return {
    header: { ...defaultTheme.header, ...header },
    rows: {
      ...defaultTheme.rows,
      ...rows,
      stripes: {
        ...defaultTheme.rows.stripes,
        ...rows?.stripes,
      },
    },
    grid: { ...defaultTheme.grid, ...grid },
  }
}

export const getFormattedTheme = (
  theme: TableQLTheme,
): TableQLProviderTheme => {
  return isBasicTheme(theme)
    ? formattedBasicTheme(theme)
    : formattedAdvancedTheme(theme)
}
