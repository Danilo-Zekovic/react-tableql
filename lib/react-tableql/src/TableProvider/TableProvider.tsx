import React, {
  createContext,
  FC,
  ReactElement,
  useContext,
  useEffect,
  useReducer,
} from 'react'

import { defaultTheme } from '../shared/defaultTheme'
import { traverseObjectForArray } from '../utils'
import { useTableQLState } from '../TableQLProvider'
import { getFormattedTheme } from './getFormattedTheme'
import { tableReducer } from './reducer'
import { State, TableProviderProps } from './types'

const getHeaderLabels = (data: object): string[] => {
  // log('Get header labels.')
  const labels = [] as string[]
  for (const key in data) {
    // exception to eliminate meta fields
    if (!key.includes('__')) {
      labels.push(key)
    }
  }

  return labels
}

const defaultState: State = { theme: defaultTheme, data: [], dataKeys: [] }

const TableStateContext = createContext<State>(defaultState)

const TableProvider: FC<TableProviderProps> = ({
  data,
  children,
  theme,
  columns,
}): ReactElement => {
  // TODO as of right now this is line makes it mandatory to have the TableQLProvider, but we do not want to force it onto users
  const { theme: globalTheme } = useTableQLState()
  const [state, dispatch] = useReducer(tableReducer, {
    ...defaultState,
    // if there is no local theme being passed use the global
    // TODO this should probably be improved so that if global is not set either just don't use it
    theme: getFormattedTheme(theme || globalTheme),
  })

  useEffect(() => {
    const traversedData: object[] = data
      ? (traverseObjectForArray(data) as object[])
      : []

    // setDisplayData(traversedData)
    // setFullData(traversedData)

    if (traversedData.length) {
      dispatch({
        type: 'setData',
        payload: { data: traversedData },
      })
    }
    dispatch({
      type: 'setDataKeys',
      payload: { dataKeys: columns || getHeaderLabels(traversedData[0]) },
    })
  }, [data, columns])

  return (
    <TableStateContext.Provider value={state}>
      {children}
    </TableStateContext.Provider>
  )
}

const useTableState = (): State => {
  const context = useContext(TableStateContext)
  /* istanbul ignore next */
  if (context === undefined) {
    throw new Error('useTableState must be used within a TableQLProvider')
  }
  return context
}

export { TableProvider, useTableState }
