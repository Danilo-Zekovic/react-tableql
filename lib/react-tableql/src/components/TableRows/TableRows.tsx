import React, { FC } from 'react'
import { Column } from '../../shared/types'

import { useTableState } from '../../TableProvider'
import { TableCell } from '../TableCell'

const TableRows: FC = () => {
  const { data, dataKeys } = useTableState()

  return (
    <>
      {data.map((row: object) => (
        <tr
          key={`TableQLRow${JSON.stringify(row)}`}
          // className={styles.tbodyTr || 'TableQL-tr'}
          className={`TableQL-tr`}
          // onClick={onRowClick ? () => onRowClick(row) : undefined}
        >
          {dataKeys.map((column: Column, columnIndex: number) => (
            <TableCell
              key={`TableCell${JSON.stringify(column)}${columnIndex}`}
              column={column}
              // styles={styles}
              // data={row[typeof column === 'string' ? column : column.id]}
              data={row}
            />
          ))}
        </tr>
      ))}
    </>
  )
  // TODO rm
  // return (
  //   <>
  //     <tr className={`TableQL-tr`}>
  //       <td className={`TableQL-td`} data-label="Account">
  //         Visa - 3412
  //       </td>
  //       <td className={`TableQL-td`} data-label="Due Date">
  //         04/01/2016
  //       </td>
  //       <td className={`TableQL-td`} data-label="Amount">
  //         $1,190
  //       </td>
  //       <td className={`TableQL-td`} data-label="Period">
  //         03/01/2016 - 03/31/2016
  //       </td>
  //     </tr>
  //     <tr className={`TableQL-tr`}>
  //       <td scope="row" className={`TableQL-td`} data-label="Account">
  //         Visa - 6076
  //       </td>
  //       <td className={`TableQL-td`} data-label="Due Date">
  //         03/01/2016
  //       </td>
  //       <td className={`TableQL-td`} data-label="Amount">
  //         $2,443
  //       </td>
  //       <td className={`TableQL-td`} data-label="Period">
  //         02/01/2016 - 02/29/2016
  //       </td>
  //     </tr>
  //     <tr className={`TableQL-tr`}>
  //       <td scope="row" className={`TableQL-td`} data-label="Account">
  //         Corporate AMEX
  //       </td>
  //       <td className={`TableQL-td`} data-label="Due Date">
  //         03/01/2016
  //       </td>
  //       <td className={`TableQL-td`} data-label="Amount">
  //         $1,181
  //       </td>
  //       <td className={`TableQL-td`} data-label="Period">
  //         02/01/2016 - 02/29/2016
  //       </td>
  //     </tr>
  //     <tr className={`TableQL-tr`}>
  //       <td scope="row" className={`TableQL-td`} data-label="Acount">
  //         Visa - 3412
  //       </td>
  //       <td className={`TableQL-td`} data-label="Due Date">
  //         02/01/2016
  //       </td>
  //       <td className={`TableQL-td`} data-label="Amount">
  //         $842
  //       </td>
  //       <td className={`TableQL-td`} data-label="Period">
  //         01/01/2016 - 01/31/2016
  //       </td>
  //     </tr>
  //   </>
  // )
}

export default TableRows
