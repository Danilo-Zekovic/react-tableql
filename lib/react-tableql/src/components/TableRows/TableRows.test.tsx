import React from 'react'
import { render } from '@testing-library/react'

import TableRows from './TableRows'
import { TableProvider } from '../../TableProvider'
import { MOCK_DATA, MOCK_DATA_KEYS } from '../../shared/testUtils'

describe('<TableRows>', () => {
  it('default snapshot', () => {
    const tbody = document.createElement('tbody')
    const { container } = render(
      <TableProvider data={MOCK_DATA}>
        <TableRows />
      </TableProvider>,

      {
        container: document.body.appendChild(tbody),
      },
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('when columns config passed', () => {
    const thead = document.createElement('thead')

    const { container } = render(
      <TableProvider
        data={MOCK_DATA}
        columns={[MOCK_DATA_KEYS[0], { id: MOCK_DATA_KEYS[1] }]}
      >
        <TableRows />
      </TableProvider>,
      {
        container: document.body.appendChild(thead),
      },
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
