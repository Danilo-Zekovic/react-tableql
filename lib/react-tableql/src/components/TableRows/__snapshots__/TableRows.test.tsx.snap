// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableRows> default snapshot 1`] = `
<tr
  class="TableQL-tr"
>
  <td
    class="TableQL-td"
    data-label="author"
  >
    Austen, Jane
  </td>
  <td
    class="TableQL-td"
    data-label="title"
  >
    Northanger Abbey
  </td>
  <td
    class="TableQL-td"
    data-label="id"
  >
    63a32284072b2e31190bd81d
  </td>
  <td
    class="TableQL-td"
    data-label="edition"
  >
    Penguin
  </td>
  <td
    class="TableQL-td"
    data-label="price"
  >
    18.2
  </td>
  <td
    class="TableQL-td"
    data-label="yearWritten"
  >
    1814
  </td>
</tr>
`;

exports[`<TableRows> when columns config passed 1`] = `
<tr
  class="TableQL-tr"
>
  <td
    class="TableQL-td"
    data-label="author"
  >
    Austen, Jane
  </td>
  <td
    class="TableQL-td"
    data-label="title"
  >
    Northanger Abbey
  </td>
</tr>
`;
